src/search/parser/lexical_analyzer.cc

```diff
@@ -0,0 +1,131 @@
+#include "lexical_analyzer.h"
+
+#include "errors.h"
+
+#include "../utils/collections.h"
+#include "../utils/system.h"
+
+#include <cassert>
+#include <regex>
+
+using namespace std;
+
+namespace parser {
+static regex build_regex(const string &s, regex::flag_type f = regex_constants::ECMAScript) {
+    return regex("^\\s*(" + s + ")\\s*", f);
+}
+
+static const vector<pair<TokenType, regex>> token_type_expressions = {
+    {TokenType::OPENING_PARENTHESIS, build_regex("\\(")},
+    {TokenType::CLOSING_PARENTHESIS, build_regex("\\)")},
+    {TokenType::OPENING_BRACKET, build_regex("\\[")},
+    {TokenType::CLOSING_BRACKET, build_regex("\\]")},
+    {TokenType::COMMA, build_regex(",")},
+    {TokenType::EQUALS, build_regex("=")},
+    {TokenType::FLOAT, build_regex(
+       R"([+-]?((\d*\.\d+|\d+\.)(e[+-]?\d+|[kmg])?)\b)",
+       regex_constants::ECMAScript | regex_constants::icase)},
+    {TokenType::INTEGER, build_regex(
+       R"([+-]?(infinity|\d+(e+?\d+|[kmg])?)\b)",
+       regex_constants::ECMAScript | regex_constants::icase)},
+    {TokenType::BOOLEAN, build_regex(R"(true|false)",
+       regex_constants::ECMAScript | regex_constants::icase)},
+    {TokenType::LET, build_regex("let",
+       regex_constants::ECMAScript | regex_constants::icase)},
+    {TokenType::IDENTIFIER, build_regex(R"([a-zA-Z_]\w*)")}
+};
+
+
+static vector<Token> split_tokens(const string &text) {
+    vector<Token> tokens;
+    auto start = text.begin();
+    auto end = text.end();
+    while (start != end) {
+        bool has_match = false;
+        smatch match;
+
+        for (const auto &type_and_expression : token_type_expressions) {
+            TokenType token_type = type_and_expression.first;
+            const regex &expression = type_and_expression.second;
+            if (regex_search(start, end, match,expression)) {
```

Comment:
```
regex_search will successfully match any subsequence [...], whereas std::regex_match will only return true if the regular expression matches the entire sequence.
```
As I understand this we have to options:
1. use `regex_search` and enforce that the match starts at the first character of the input
2. use `regex_match` and enforce that the match also includes everything after a token.

Reg 1: We did the first by appending to every regular expression `^` which means the regex matches only if the match happens at the first char. This has also the advantage that the size of the match is exactly the token.
Reg 2: We could do this by adding a catch all `.*` to the regex, but then we have to extract the relevant token from the matched sequence (via matching groups?)

Reply:
I see, that's different from how I know it in other programming languages then where "match" anchors to the start, but not to the end (i.e., it matches any prefix) For example: https://docs.python.org/3/library/re.html#re.Pattern.match
Apparently Python added `fullmatch` to match the whole string in Python 3.4, which I didn't previously know.

I think anchoring with `^` makes sense if that works in conjunction with the start/end argument. Rather than adding `^` manually everywhere ourselves, this could be part of an initialization loop.

Reply:
Hmmm, seems to be a bit more complicated regarding interaction with newlines: https://stackoverflow.com/questions/39645660/stdregex-to-match-begin-end-of-string
Do we/should we split the string into lines and process them separately? This is often done anyway because it helps do things like report on which line an error was. We could also combine everything into a single line by replacing '\n' with a space before further processing, but I'm not sure I like this better. There is probably nothing wrong with people wanting to split complex specifications into multiple lines.

Reply:
Ok, we already sanitize the arguments in `commandline.cc`. This sanitization does:
- replace `\n` with ` `
- convert to lower case
Simplest would be do just keep this sanitization.

If we want to report the lexer errors on a per line basis, then we could:
- We could make the search argument a special case which does only the conversion to lower case, but I dislike this special case.
- stop replacing `\n` by ` `. Then the user has to take care to use `\n` only where is is allowed (or we add a check for this). It is strange that we accept a plan file path with `\n` and then store it to the path where we replace `\n` by ` `.

Reply:
I think the sanitization was primarily done for the benefit of the old option parser. Is it still done for all arguments, even ones unrelated to the option parser? I thought we changed this at some point because people complained that file names that were passed in were converted to lower case. In general, I think the sanitization only makes sense for the option parser, and if we want to do it, we should do it inside the lexical analysis rather than inside the calling code. If I recall correctly (which I may not), out reg-exes already include options like "ignore case" which suggest they are intended for possibly mixed-case inputs.

Reply:
BTW, we could also consider getting rid of the case-insensitivity of the option parser at some point. Maybe not right now because right now we're not introducing backwards-incompatible changes to the command-line and it's good not to do this too often, but perhaps at the point where we get rid of the old `--evaluator` etc. options. If we intend to print deprecation warnings about these at some point (which I don't think is necessary, but we might), then at this point we could also print deprecation warnings for option parser string that contain upper-case characters.

Reply:
We now have the conversion to lower case in the lexical analysis. Currently, we do this for all tokens, because none of our token types is case sensitive. In the future, we could for example introduce a token type for a case sensitive string literal and then skip the conversion to lower case there.

I'll mark this conversation as resolved but feel free to reopen if you disagree.